scripts/test_solidity_lsp.py

```diff
@@ -0,0 +1,379 @@
+#!/usr/bin/env python3.8
+
+import argparse
+import json
+import os
+import subprocess
+
+from typing import List, Union, Any
+from deepdiff import DeepDiff
+
+# {{{ JsonRpcProcess
+class MyEncoder(json.JSONEncoder):
+    """
+    Encodes an object in JSON
+    """
+    def default(self, o): # pylint: disable=E0202
+        return o.__dict__
+
+class JsonRpcProcess:
+    exe_path: str
+    exe_args: List[str]
+    process: subprocess.Popen
+    trace_io: bool
+
+    def __init__(self, exe_path: str, exe_args: List[str], trace_io: bool = False):
+        self.exe_path = exe_path
+        self.exe_args = exe_args
+        self.trace_io = trace_io
+
+    def __enter__(self):
+        self.process = subprocess.Popen(
+            [self.exe_path, *self.exe_args],
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE
+        )
+        return self
+
+    def __exit__(self, exception_type, exception_value, traceback) -> None:
+        self.process.kill()
+        self.process.wait(timeout=2.0)
+
+    def trace(self, topic: str, message: str) -> None:
+        if self.trace_io:
+            print(f"{SGR_TRACE}{topic}:{SGR_RESET} {message}")
+
+    def receive_message(self) -> Any:
+        # `, timeout: float = 2.0`
+        # Note, we should make use of timeout to avoid infinite blocking if nothing is received.
+        LEN_HEADER = "Content-Length: "
+        TYPE_HEADER = "Content-Type: "
+        if self.process.stdout == None:
+            return None
+        message_size = None
+        while True:
+            #read header
+            line = self.process.stdout.readline()
+            if not line:
+                # server quit
+                return None
+            line = line.decode("utf-8")
+            if not line.endswith("\r\n"):
+                raise RuntimeError("Bad header: missing newline")
```

Comment:
I'd recommend defining some more specific exception types:
```suggestion
                raise BadHeader("Bad header: missing newline")
```